Library/Homebrew/download_strategy.rb

```diff
@@ -282,6 +283,14 @@ def initialize(name, resource)
   end
 
   def fetch
+    if ENV["HOMEBREW_ARTIFACT_DOMAIN"]
+      if @url.start_with?(BottleSpecification::DEFAULT_DOMAIN)
+        @url.sub!(BottleSpecification::DEFAULT_DOMAIN, ENV["HOMEBREW_ARTIFACT_DOMAIN"])
```

Comment:
IMHO, I think `HOMEBREW_ARTIFACT_DOMAIN` should work in the way to be oblivious of bottles. My defenses are following:

* It's unexpected to let `HOMEBREW_ARTIFACT_DOMAIN` overwrite `HOMEBREW_BOTTLE_DOMAIN`.
* If users want to use artifact services other than JForg Aritfactory(e.g. self hosted in enterprise settings). They should get it without any special settings. This means that bottle should not get special treatment. More specially the download urls would be like these:
  * For bottles:  `https://example.com/artifactory/homebrew/homebrew.bintray.com/bottles/gcc-6.1.0.el_capitan.bottle.tar.gz`
   * For non bottles: `https://example.com/artifactory/homebrew/s3.amazonaws.com/downloads.basho.com/riak/2.1/2.1.4/osx/10.8/riak-2.1.4-OSX-x86_64.tar.gz`

> Handling bottles are easy as its coming from only one source which is https://homebrew.bintray.com/bottles

* This is not correct. We have different sources for different taps.
* Even with full url like `https://example.com/artifactory/homebrew/homebrew.bintray.com/bottles/gcc-6.1.0.el_capitan.bottle.tar.gz`. You can still easily handle bottles fetching. The only different is the url prefix is different. 

My main intention is we should support such feature in a general fashion. So any artifact services can be applied to this. At the same time, users don't have to know anything about bottles to make it work.